docs/src/sections/webapp_solution.adoc

```diff
@@ -0,0 +1,529 @@
+że token jest w trakcie ładowania.
+
+.NavbarLayout.tsx.
+[source, tsx]
+----
+include::../../../webapp/src/components/Layouts/NavbarLayout.tsx[lines=9..32]
+----
+
+.Wiadomość o wczytywaniu tokenu w lewym dolnym rogu.
+image::../images/system_presentation/webapp/refreshing_token.png[]
+
+_UserLayout.tsx_, tak samo jak _AdminLayout.tsx_ i _PublicLayout.tsx_ przekierowuje użytkownika
+w zależności od jego stanu.
+Zagnieżdżają w sobie komponenty w zależności od tego, komu należy się udostępnić dany widok.
+
+* _UserLayout.tsx_ — przekierowuje do strony logowania, jeśli użytkownik nie jest zalogowany, lub do strony głównej
+, jeśli użytkownik ma rolę tymczasową
+* _AdminLayout.tsx_ — przekierowuje do strony logowania, jeśli użytkownik nie jest zalogowany, lub do strony głównej
+, jeśli użytkownik nie jest adminem
+* _PublicLayout.tsx_ — przekierowuje do panelu użytkownika, jeśli użytkownik jest zalogowany
+i ma rolę inną niż tymczasową
+
+.UserLayout.tsx.
+[source, tsx]
+----
+include::../../../webapp/src/components/Layouts/UserLayout.tsx[lines=9..18]
+----
+
+_UserContext.tsx_ jest odpowiedzialny za przechowywanie danych o użytkowniku oraz sprawdzanie ważności tokenu dostępu
+do serwera. Jeśli odświeżanie tokenu się nie powiedzie, użytkownik jest wylogowany, a zarazem przekierowany
+przez _UserLayout.tsx_ lub _AdminLayout.tsx_ do strony logowania.
+Jeśli odświeżanie się powiedzie, kolejne odświeżenie zostaje zakolejkowane do wykonania za 10 minut.
+Ponieważ ważność tokenu na serwerze jest ustalona na 15 minut, token w przeglądarce jest zawsze ważny.
+Udostępnia również pomocnicze metody w celu wyeliminowania powtórzeń np. _login_, _logout_.
+
+.Funkcje _login_ i _logout_ pliku UserContext.tsx.
+[source, tsx]
+----
+include::../../../webapp/src/components/Context/UserContext.tsx[lines=47..59]
+----
+
+.Odświeżanie tokenu w pliku UserContext.tsx.
+[source, tsx]
+----
+include::../../../webapp/src/components/Context/UserContext.tsx[lines=61..79]
+----
+
+Conext _UserContext.tsx_ obejmuje wszystkie widoki,
+tak żeby dane o użytkowniku nie zostały usunięte przy zmianie widoków.
+Do tego korzysta z pliku pomocniczego _src/utils/persist.ts_,
+aby dane o użytkowniku zostały zapisane nawet przy zamknięciu przeglądarki.
+
+Plik _persist.ts_ zawiera funkcje dotyczące _localStorage_,
+czyli API do długotrwałego zapisu danych dostępnego w przeglądarce.
+API _localStorage_ pozwala zapisać dane typu klucz wartość, ale klucz i wartość muszą być wartością tekstową.
+Zostały napisane funkcje, aby obejść te ograniczenie, zamieniając obiekty na wartość tekstową przy zapisywaniu
+i odwrotnie przy wczytywaniu.
+
+Funkcja _setLocalJson_ zamienia przyjętą wartość w ciąg tekstowy i zapisuje go w _localStorage_.
+
+.Funkcja setLocalJson w pliku persist.ts.
+[source, typescript]
+----
+include::../../../webapp/src/utils/persist.ts[lines=13..15]
+----
+
+Funkcja _getLocalJson_ pobiera wartość z _localStorage_ i zamienia ją z ciągu tekstowego na _JSON_,
```

Comment:
```suggestion
Funkcja _getLocalJson_ pobiera wartość z _localStorage_ i zamienia ją z ciągu tekstowego na obiekt_JSON_,
```

Reply:
JSON to JavaScript Object Notation, to nie będzie trochę masło maślane?